### ISSUE TITLE
"github.com/golang/protobuf/protoc-gen-go/descriptor" is provided by more than one rule

### ISSUE THREAD

**Issue by linzhp** [bazelbuild/rules_go#2537]
`@io_bazel_rules_go//proto:go_proto` depends on both`@io_bazel_rules_go//proto/wkt:descriptor_go_proto` and `@com_github_golang_protobuf//protoc-gen-go/descriptor:go_default_library`, which can be seen using:

```bash
bazel query 'allpaths(@io_bazel_rules_go//proto:go_proto, @com_github_golang_protobuf//protoc-gen-go/descriptor:go_default_library)' --notool_deps --output graph | dot -Tsvg > /tmp/deps.svg
```

Both of them have "github.com/golang/protobuf/protoc-gen-go/descriptor" as `importpath`. So when linking any `go_binary` that depends on `@io_bazel_rules_go//proto:go_proto`, there will be warnings like:

```
link: warning: package "github.com/golang/protobuf/protoc-gen-go/descriptor" is provided by more than one rule:
 @io_bazel_rules_go//proto/wkt:descriptor_go_proto
 @com_github_golang_protobuf//protoc-gen-go/descriptor:go_default_library
```

### What version of rules_go are you using?

v0.23.3

### What version of gazelle are you using?

47a2afcc7eb6e7b5dca4a791a77f459738a20483

### What version of Bazel are you using?

3.1

### Does this issue reproduce with the latest releases of all the above?

Yes

### What operating system and processor architecture are you using?

Linux and Darwin, amd64

### What did you do?

Build this file with Bazel:

```go
package main

import (
	"fmt"

	_ "google.golang.org/grpc"
)

func main() {
	fmt.Println("Hello, world")
}
```

### What did you expect to see?

Build pass without warning

### What did you see instead?

```
link: warning: package "github.com/golang/protobuf/protoc-gen-go/descriptor" is provided by more than one rule:
    @io_bazel_rules_go//proto/wkt:descriptor_go_proto
    @com_github_golang_protobuf//protoc-gen-go/descriptor:go_default_library
Set "importmap" to different paths in each library.
This will be an error in the future.
```

**Comment by jayconrod** [comment]
Depending on `@io_bazel_rules_go//proto/wkt:descriptor_go_proto` shouldn't cause this error. It depends on `@com_github_golang_protobuf//protoc-gen-go/descriptor:go_default_library` in a way that only one package gets passed to the linker. The error indicates that two or more packages were passed to the linker with the same `importpath`.

Could you confirm using `bazel query` or something similar that nothing else directly depends on `@com_github_golang_protobuf//protoc-gen-go/descriptor:go_default_library`?

If there are no other dependencies, could you please post a small repro for this?

**Comment by linzhp** [comment]
Updated the description of the issue with a small repro, and the `bazel query` command to show two paths going from `@io_bazel_rules_go//proto:go_proto`  to `@com_github_golang_protobuf//protoc-gen-go/descriptor:go_default_library`

**Comment by linzhp** [comment]
It turned out `@io_bazel_rules_go//proto:go_proto` depends on `@com_github_golang_protobuf//protoc-gen-go/plugin:go_default_library` and at v1.3.3, `@com_github_golang_protobuf//protoc-gen-go/plugin:go_default_library` depends on `@com_github_golang_protobuf//protoc-gen-go/descriptor:go_default_library`. Upgrading `github.com/golang/protobuf` to v1.4.1 fixed the issue.

**Comment by linzhp** [comment]
It turned out that `github.com/golang/protobuf@v1.4` an implementation of [APIv2](https://blog.golang.org/protobuf-apiv2), which has breaking changes. The upgrade is not as straightforward as I hoped. @jayconrod Do you think it worth fixing the linker warning for `github.com/golang/protobuf@v1.3`?

**Comment by jayconrod** [comment]
I think when I read this earlier, I assumed you were using the `com_github_golang_protobuf` declared in `go_rules_dependencies`. In rules_go `v0.23.3`, that's `v1.4.1`. I can't support older versions on this branch, since we're starting to make assumptions about APIv2. protobuf `v1.3.3` may still work with rules_go `v0.22.x` though.

> Do you think it worth fixing the linker warning for github.com/golang/protobuf@v1.3?

Not sure what you mean here. The linker warning indicates that two packages are passed to the linker with the same package path. This will be an error later on.

In rules_go `v0.23.3`, `@io_bazel_rules_go//proto/wkt:descriptor_go_proto` is a wrapper for `@com_github_golang_protobuf//protoc-gen-go/descriptor:go_default_library`; depending on ``@io_bazel_rules_go//proto/wkt:descriptor_go_proto` alone should not cause two packages with that path to be passed to the linker. There may be some other chain of dependencies to `@com_github_golang_protobuf//protoc-gen-go/descriptor:go_default_library`.

**Comment by linzhp** [comment]
Yeah, we override `com_github_golang_protobuf` in `go_rules_dependencies ` with the one `go mod` resolves for us, so we are still in v1.3.3. With rules_go v0.23.3 and `com_github_golang_protobuf` v1.3.3, the dependency graph (with `--notool_deps`) looks like the following:

![deps](https://user-images.githubusercontent.com/98395/85900484-c24c5480-b7b4-11ea-8b30-3ce71becd27f.jpg)

The left path exist in `com_github_golang_protobuf` v1.3.3, but not in v1.4.1. The right path was introduced to rules_go in 740ada94dfda52f2a079f718858e8b2b8ee0fdc6. So it is unfortunate for those who stuck at `com_github_golang_protobuf` v1.3.3 and use newer version of rules_go that includes 740ada94dfda52f2a079f718858e8b2b8ee0fdc6 to have both paths.

**Comment by jayconrod** [comment]
Thanks for explaining. The graph makes it clear what's happening.

The best thing to do is probably to downgrade to rules_go v0.22.x or upgrade to protobuf v1.4.x. A single version of rules_go cannot easily support new and old versions of protobuf.

In protobuf v1.4.x, the code in `github.com/golang/protobuf/protoc-gen-go/plugin` is a wrapper for the equivalent APIv2 package, `google.golang.org/protobuf/types/pluginpb`. The other WKTs are similar. They are generated using some special tooling that rules_go does not have access to from the command line. That means `@io_bazel_rules_go//proto/wkt:compiler_plugin_go_proto` can't easily generate the same code using `protoc-gen-go`. I mistakenly tried to do that in `v0.23.0`, but it caused conflicts (#2501).

To fix that conflict, the protobuf team advised me to find a way to use the pre-generated code. So all of the `@io_bazel_rules_go//go/wkt:*_go_proto` targets use a fake compiler, `go_proto_wrapper`, which is compatible with `go_proto_compiler` but returns pregenerated sources instead of invoking `protoc-gen-go`. That works fine for protobuf `v1.4.x`, but not for anything before that.

Closing this issue because I don't think a fix in rules_go is feasible.
